File: Source/Remote Control Server/ApiV1.vb
Imports System.Windows.Forms

Module ApiV1

    Public Const cmd_info_device_name As String = "[cmd_15]"
    Public Const cmd_info_device_osversion As String = "[cmd_16]"
    Public Const cmd_info_app_name As String = "[cmd_17]"
    Public Const cmd_info_app_version As String = "[cmd_18]"
    Public Const cmd_info_app_mode As String = "[cmd_19]"


    Public Const cmd_connect As String = "[cmd_connect]"
    Public Const cmd_disconnect As String = "[cmd_disconnect]"
    Public Const cmd_lizens As String = "[cmd_lizens]"
    Public Const cmd_version As String = "[cmd_version]"

    Public Const cmd_mouse_string As String = "[cmd_mouse]"
    Public Const cmd_keyboard_string As String = "[cmd_keyboard]"
    Public Const cmd_media_string As String = "[cmd_music]"
    Public Const cmd_slideshow_string As String = "[cmd_slide]"
    Public Const cmd_process_string As String = "[cmd_process]"
    Public Const cmd_scroll_string As String = "[cmd_scroll]"
    Public Const cmd_broadcast_string As String = "[cmd_broadcast]"
    Public Const cmd_google_string As String = "[cmd_google]"
    Public Const cmd_shortcut_string As String = "[cmd_short]"

    Public readableCommand As String = "Unknown"

    Public Sub parseCommand(ByRef command As Command)
        Dim cmd As String = command.dataAsString
        Dim value As String = getCommandValue(cmd)
        Dim app As App = Server.getApp(command.source)

        readableCommand = "Unknown"
        If cmd.Contains(cmd_mouse_string) Then
            parseMouseCommand(value)
            Logger.add("Mouse: " & value)
        ElseIf cmd.Contains(cmd_keyboard_string) Then
            parseKeyboardCommand(cmd)
            Logger.add("Keyboard: " & readableCommand)
        ElseIf cmd.Contains(cmd_media_string) Then
            If (app.license.isProVersion) Then
                parseMediaCommand(cmd)
            Else
                app.license.requestUpgrade()
            End If
            Logger.add("Media: " & readableCommand)
        ElseIf cmd.Contains(cmd_slideshow_string) Then
            If (app.license.isProVersion) Then
                parseSlideshowCommand(cmd)
            Else
                app.license.requestUpgrade()
            End If
            Logger.add("Slideshow: " & readableCommand)
        ElseIf cmd.Contains(cmd_scroll_string) Then
            parseScrollCommand(cmd)
            Logger.add("Scroll: " & readableCommand)
        ElseIf cmd.Contains(cmd_shortcut_string) Then
            parseShortcutCommand(cmd)
            Logger.add("Shortcut: " & readableCommand)
        ElseIf cmd.Contains(cmd_process_string) Then
            Try
                Process.Start(value)
            Catch ex As Exception
            End Try
            Logger.add("Process: " & value)
        ElseIf cmd.Contains(cmd_google_string) Then
            Try
                Process.Start("http://google.de/search?q=" & value)
            Catch ex As Exception
            End Try
            Logger.add("Google: " & value)

            'Connection
        ElseIf cmd.Contains(cmd_broadcast_string) Then
            command.source = value
            app = Server.getApp(value)
            app.onBroadCast(command)
        ElseIf cmd.Contains(cmd_connect) Then
            app.deviceName = value
            Logger.add("Device name set to " & value)
            app.onConnect()
        ElseIf cmd.Contains(cmd_disconnect) Then
            app = Server.getApp(value)
            app.onDisconnect()

            'Info
        ElseIf cmd.Contains(cmd_info_device_name) Then
            app.deviceName = value
            Logger.add("Device name set to " & value)
        ElseIf cmd.Contains(cmd_info_device_osversion) Then
            app.osVersion = value
            app.detectOs()
            Logger.add("Device OS version set to " & value)
        ElseIf cmd.Contains(cmd_info_app_name) Then
            app.appName = value
            Logger.add("App name set to " & value)
        ElseIf cmd.Contains(cmd_info_app_version) Or cmd.Contains(cmd_version) Then
            app.appVersion = value
            Logger.add("App version set to " & value)
        ElseIf cmd.Contains(cmd_info_app_mode) Or cmd.Contains(cmd_lizens) Then
            Dim license As New License
            If value.ToLower.Equals("pro") Or value.ToLower.Equals("trial") Then
                license.isProVersion = True
            Else
                license.isProVersion = False
            End If
            app.license = license
            Logger.add("App license set to " & value)
        Else
            Logger.add("Unknown ApiV1: " + cmd)
        End If

    End Sub

    Public Function getCommandValue(ByVal cmd As String) As String
        Dim value As String
        value = cmd.Substring(InStr(cmd, "]"))
        value = value.Replace(vbLf, "")
        Return value
    End Function

    Public Sub parseMouseCommand(ByVal cmd As String)
        Mouse.parseMouse(cmd)
    End Sub

    Public Sub parseKeyboardCommand(ByVal cmd As String)
        If cmd.Contains("<") And cmd.Contains(">") Then
            If cmd.Contains("<Enter>") Then
                readableCommand = "Enter"
                Keyboard.sendKeyPress(Keys.Enter)
                Exit Sub
            ElseIf cmd.Contains("<Back>") Then
                readableCommand = "Backspace"
                Keyboard.sendKeyPress(Keys.Back)
                Exit Sub
            ElseIf cmd.Contains("<tab>") Then
                readableCommand = "Tab"
                Keyboard.sendKeyPress(Keys.Tab)
                Exit Sub
            ElseIf cmd.Contains("<caps>") Then
                readableCommand = "Capslock"
                Keyboard.sendKeyPress(Keys.CapsLock)
                Exit Sub
            ElseIf cmd.Contains("<ctrl>") Then
                readableCommand = "Control"
                If cmd.Contains("down") Then
                    sendKeyDown(Keys.ControlKey)
                Else
                    sendKeyUp(Keys.ControlKey)
                End If
                Exit Sub
            ElseIf cmd.Contains("<alt>") Then
                readableCommand = "Alt"
                If cmd.Contains("down") Then
                    'SendKeyDown(Keys.Alt)
                    sendKeyDown(&H1)
                Else
                    'SendKeyUp(Keys.Alt)
                    sendKeyUp(&H1)
                End If
                Exit Sub
            ElseIf cmd.Contains("<shift>") Then
                readableCommand = "Shift"
                If cmd.Contains("down") Then
                    sendKeyDown(Keys.ShiftKey)
                Else
                    sendKeyUp(Keys.ShiftKey)
                End If
            ElseIf cmd.Contains("<del>") Then
                readableCommand = "Delete"
                Keyboard.sendKeyPress(Keys.Delete)
            ElseIf cmd.Contains("<win>") Then
                readableCommand = "Windows"
                Keyboard.sendKeyPress(Keys.LWin)
                Exit Sub
            ElseIf cmd.Contains("<esc>") Then
                readableCommand = "Escape"
                Keyboard.sendKeyPress(Keys.Escape)
                Exit Sub
            ElseIf cmd.Contains("<end>") Then
                readableCommand = "End"
                Keyboard.sendKeyPress(Keys.End)
                Exit Sub
            ElseIf cmd.Contains("<tab>") Then
                readableCommand = "Tab"
                Keyboard.sendKeyPress(Keys.Tab)
                Exit Sub
            ElseIf cmd.Contains("<ins>") Then
                readableCommand = "Insert"
                Keyboard.sendKeyPress(Keys.Insert)
                Exit Sub
            ElseIf cmd.Contains("<up>") Then
                readableCommand = "Up"
                Keyboard.sendKeyPress(Keys.Up)
                Exit Sub
            ElseIf cmd.Contains("<down>") Then
                readableCommand = "Down"
                Keyboard.sendKeyPress(Keys.Down)
                Exit Sub
            ElseIf cmd.Contains("<left>") Then
                readableCommand = "Left"
                Keyboard.sendKeyPress(Keys.Left)
                Exit Sub
            ElseIf cmd.Contains("<right>") Then
                readableCommand = "Right"
                Keyboard.sendKeyPress(Keys.Right)
                Exit Sub
            End If
        Else
            Try
                readableCommand = getCommandValue(cmd)
                Keyboard.sendEachKey(readableCommand)
            Catch ex As Exception
                Logger.add(ex.ToString)
            End Try
        End If
    End Sub

    Public Sub parseMediaCommand(ByVal cmd As String)
        Dim value As String = getCommandValue(cmd)
        If value = "play" Then
            readableCommand = "Play"
            Media.playMedia()
        ElseIf value = "stop" Then
            readableCommand = "Stop"
            Media.stopMedia()
        ElseIf value = "next" Then
            readableCommand = "Next"
            Media.nextMedia()
        ElseIf value = "prev" Then
            readableCommand = "Previous"
            Media.previousMedia()
        ElseIf value = "volup" Then
            readableCommand = "Volume up"
            Media.volumeUp()
        ElseIf value = "voldown" Then
            readableCommand = "Volume down"
            Media.volumeDown()
        ElseIf value = "mute" Then
            readableCommand = "Mute"
            Media.volumeMute()
        ElseIf value = "launch" Then
            readableCommand = "Launch player"
            Media.launchPlayer()
        End If
    End Sub

    Public Sub parseScrollCommand(ByVal cmd As String)
        Dim value As String = getCommandValue(cmd)
        If value = "zoomin" Then
            readableCommand = "Zoom in"
            Mouse.zoom(1, 1)
        ElseIf value = "zoomout" Then
            readableCommand = "Zoom out"
            Mouse.zoom(-1, 1)
        ElseIf value = "back" Then
            readableCommand = "Back"
            Keyboard.sendKeyPress(Keys.BrowserBack)
        ElseIf value = "forward" Then
            readableCommand = "Forward"
            Keyboard.sendKeyPress(Keys.BrowserForward)
        ElseIf value = "pageup" Then
            readableCommand = "Page up"
            Keyboard.sendKeyPress(Keys.PageUp)
        ElseIf value = "pagedown" Then
            readableCommand = "Page down"
            Keyboard.sendKeyPress(Keys.PageDown)
        ElseIf value = "cancel" Then
            readableCommand = "Cancel"
            Keyboard.sendKeyPress(Keys.BrowserStop)
        ElseIf value = "refresh" Then
            readableCommand = "Refresh"
            Keyboard.sendKeyPress(Keys.BrowserRefresh)
        ElseIf value = "fullexit" Then
            readableCommand = "Exit fullscreen"
            Keyboard.sendKeyPress(Keys.Escape)
        ElseIf value = "fullscreen" Then
            readableCommand = "Fullscreen"
            Keyboard.sendKeyPress(Keys.F11)
        End If
    End Sub

    Public Sub parseShortcutCommand(ByVal cmd As String)
        Dim value As String = getCommandValue(cmd)
        If value = "desktop" Then
            readableCommand = "Show desktop"
            Keyboard.sendShortcut(New List(Of Key) From {Keys.LWin, Keys.D})
        ElseIf value = "close" Then
            readableCommand = "Close"
            'Doesn't work in the .Net Framework:
            'Keyboard.sendShortcut(New List(Of Key) From {Keys.Alt, Keys.F4})
            Keyboard.sendKeys("%{F4}")
        ElseIf value = "copy" Then
            readableCommand = "Copy"
            Keyboard.sendShortcut(New List(Of Key) From {Keys.ControlKey, Keys.C})
        ElseIf value = "paste" Then
            readableCommand = "Paste"
            Keyboard.sendShortcut(New List(Of Key) From {Keys.ControlKey, Keys.V})
        ElseIf value = "selectall" Then
            readableCommand = "Select all"
            Keyboard.sendShortcut(New List(Of Key) From {Keys.ControlKey, Keys.A})
        ElseIf value = "undo" Then
            readableCommand = "Undo"
            Keyboard.sendShortcut(New List(Of Key) From {Keys.ControlKey, Keys.Z})
        ElseIf value = "standby" Then
            readableCommand = "Standby"
            Keyboard.standby()
        ElseIf value = "shutdown" Then
            readableCommand = "Shutdown"
            Keyboard.shutdown()
        End If
    End Sub

    Public Sub parseSlideshowCommand(ByVal cmd As String)
        Dim value As String = getCommandValue(cmd)
        If value = "pause" Then
            readableCommand = "Pause slideshow"
            Keyboard.sendKeyPress(Keys.B)
        ElseIf value = "next" Then
            readableCommand = "Next slide"
            Keyboard.sendKeyPress(Keys.Right)
        ElseIf value = "prev" Then
            readableCommand = "Previous slide"
            Keyboard.sendKeyPress(Keys.Left)
        End If
    End Sub

End Module

File: Source/MacOS/RemoteControlServer/apiv1.cpp
#include "apiv1.h"
#include "app.h"
#include "logger.h"
#include "media.h"
#include "mousev2.h"
#include "serial.h"
#include "server.h"
#include "settings.h"

#ifdef Q_OS_MAC
    #include "keyboardmac.h"
    #include "mousemac.h"

    #include <Carbon/Carbon.h>
#endif

#include <QDesktopServices>

#include <QDebug>

ApiV1* ApiV1::instance = NULL;

ApiV1* ApiV1::Instance()
{
    if (!instance)
    {
        instance = new ApiV1();
    }
    return instance;
}

ApiV1::ApiV1():
    cmd_info_device_name("[cmd_15]"),
    cmd_info_device_osversion("[cmd_16]"),
    cmd_info_app_name("[cmd_17]"),
    cmd_info_app_version("[cmd_18]"),
    cmd_info_app_mode("[cmd_19]"),
    cmd_connect("[cmd_connect]"),
    cmd_disconnect("[cmd_disconnect]"),
    cmd_lizens("[cmd_lizens]"),
    cmd_version("[cmd_version]"),
    cmd_mouse_string("[cmd_mouse]"),
    cmd_keyboard_string("[cmd_keyboard]"),
    cmd_media_string("[cmd_music]"),
    cmd_slideshow_string("[cmd_slide]"),
    cmd_process_string("[cmd_process]"),
    cmd_scroll_string("[cmd_scroll]"),
    cmd_broadcast_string("[cmd_broadcast]"),
    cmd_google_string("[cmd_google]"),
    cmd_shortcut_string("[cmd_short]")
{
    readableCommand = "Unknown";
}

bool ApiV1::isBroadcast(Command &command)
{
    return (command.dataAsString().contains(cmd_broadcast_string)) ? true : false;
}

void ApiV1::parseCommand(Command &command)
{
    QString cmd = command.dataAsString();
    QString value = getCommandValue(cmd);
    App *app = Server::Instance()->getApp(command.source);

    readableCommand = "Unknown";
    if (cmd.contains(cmd_keyboard_string))
    {
        parseKeyboardCommand(cmd);
        Logger::Instance()->add("Keyboard: " + readableCommand);
    }
    else if (cmd.contains(cmd_media_string))
    {
        parseMediaCommand(cmd);
        Logger::Instance()->add("Media: " + readableCommand);
    }
    else if (cmd.contains(cmd_slideshow_string))
    {
        parseSlideshowCommand(cmd);
        Logger::Instance()->add("Slideshow: " + readableCommand);
    }
    else if (cmd.contains(cmd_scroll_string))
    {
        parseScrollCommand(cmd);
        Logger::Instance()->add("Scroll: " + readableCommand);
    }
    else if (cmd.contains(cmd_shortcut_string))
    {
        parseShortcutCommand(cmd);
        Logger::Instance()->add("Shortcut: " + readableCommand);
    }
    else if (cmd.contains(cmd_process_string))
    {
        Server::Instance()->startProcess(value);
        Logger::Instance()->add("Process: " + value);
    }
    else if (cmd.contains(cmd_google_string))
    {
        QDesktopServices::openUrl("http://google.com/search?q=" + value);
        Logger::Instance()->add("Google: " + value);
    }
    else if (cmd.contains(cmd_broadcast_string))
    {
        command.source = value;
        app = Server::Instance()->getApp(value);
        app->onBroadCast(command);
    }
    else if (cmd.contains(cmd_connect))
    {
        app->deviceName = value;
        Logger::Instance()->add("Device name set to " + value);
        app->onConnect();
    }
    else if (cmd.contains(cmd_disconnect))
    {
        app = Server::Instance()->getApp(value);
        app->onDisconnect();
    }
    else if (cmd.contains(cmd_info_device_name))
    {
        app->deviceName = value;
        Logger::Instance()->add("Device name set to " + value);
    }
    else if (cmd.contains(cmd_info_device_osversion))
    {
        app->osVersion = value;
        Logger::Instance()->add("Device OS version set to " + value);
    }
    else if (cmd.contains(cmd_info_app_name))
    {
        app->appName = value;
        Logger::Instance()->add("App name set to " + value);
    }
    else if (cmd.contains(cmd_info_app_version) || cmd.contains(cmd_version))
    {
        app->appVersion = value;
        Logger::Instance()->add("App version set to " + value);
    }
    else
    {
        Logger::Instance()->add("Unknown ApiV1: " + cmd);
    }
}

QString ApiV1::getCommandValue(QString cmd)
{
    return cmd.right(cmd.length() - cmd.indexOf("]") - 1);
}

//void ApiV1::parseMouseCommand(QString cmd)
//{
//    MouseV2::Instance()->parseMouse(cmd);
//}

void ApiV1::parseKeyboardCommand(QString cmd)
{
    if (cmd.contains("<") && cmd.contains(">"))
    {
        if (cmd.contains("<Enter>"))
        {
            readableCommand = "Enter";
#ifdef Q_OS_MAC
            KeyboardMac::Instance()->sendKeyPress(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_ENTER));
#endif
            return;
        }
        else if (cmd.contains("<Back>"))
        {
            readableCommand = "Back";
#ifdef Q_OS_MAC
            KeyboardMac::Instance()->sendKeyPress(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_BACK));
#endif
            return;
        }
        else if (cmd.contains("<tab>"))
        {
            readableCommand = "Tab";
#ifdef Q_OS_MAC
            KeyboardMac::Instance()->sendKeyPress(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_TAB));
#endif
            return;
        }
        else if (cmd.contains("<caps>"))
        {
            readableCommand = "Capslock";
#ifdef Q_OS_MAC
            KeyboardMac::Instance()->sendKeyPress(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_CAPS_LOCK));
#endif
            return;
        }
        else if (cmd.contains("<ctrl>"))
        {
            readableCommand = "Control";
            if (cmd.contains("down"))
            {
#ifdef Q_OS_MAC
                KeyboardMac::Instance()->sendKeyDown(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_CONTROL));
#endif
            }
            else
            {
#ifdef Q_OS_MAC
                KeyboardMac::Instance()->sendKeyUp(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_CONTROL));
#endif
            }
            return;
        }
        else if (cmd.contains("<alt>"))
        {
            readableCommand = "Alt";
            if (cmd.contains("down"))
            {
#ifdef Q_OS_MAC
                KeyboardMac::Instance()->sendKeyDown(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_ALT));
#endif
            }
            else
            {
#ifdef Q_OS_MAC
                KeyboardMac::Instance()->sendKeyUp(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_ALT));
#endif
            }
            return;
        }
        else if (cmd.contains("<shift>"))
        {
            readableCommand = "Shift";
            if (cmd.contains("down"))
            {
#ifdef Q_OS_MAC
                KeyboardMac::Instance()->sendKeyDown(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_SHIFT));
#endif
            }
            else
            {
#ifdef Q_OS_MAC
                KeyboardMac::Instance()->sendKeyUp(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_SHIFT));
#endif
            }
            return;
        }
        else if (cmd.contains("<del>"))
        {
            readableCommand = "Delete";
#ifdef Q_OS_MAC
            KeyboardMac::Instance()->sendKeyPress(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_DEL));
#endif
            return;
        }
        else if (cmd.contains("<win>"))
        {
            readableCommand = "Windows";
#ifdef Q_OS_MAC
            KeyboardMac::Instance()->sendKeyPress(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_WINDOWS));
#endif
            return;
        }
        else if (cmd.contains("<esc>"))
        {
            readableCommand = "Escape";
#ifdef Q_OS_MAC
            KeyboardMac::Instance()->sendKeyPress(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_ESCAPE));
#endif
            return;
        }
        else if (cmd.contains("<end>"))
        {
            readableCommand = "End";
#ifdef Q_OS_MAC
            KeyboardMac::Instance()->sendKeyPress(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_MOVE_END));
#endif
            return;
        }
        else if (cmd.contains("<ins>"))
        {
            readableCommand = "Insert";
#ifdef Q_OS_MAC
            KeyboardMac::Instance()->sendKeyPress(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_INSERT));
#endif
            return;
        }
        else if (cmd.contains("<up>"))
        {
            readableCommand = "Up";
#ifdef Q_OS_MAC
            KeyboardMac::Instance()->sendKeyPress(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_UP));
#endif
            return;
        }
        else if (cmd.contains("<down>"))
        {
            readableCommand = "Down";
#ifdef Q_OS_MAC
            KeyboardMac::Instance()->sendKeyPress(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_DOWN));
#endif
            return;
        }
        else if (cmd.contains("<left>"))
        {
            readableCommand = "Left";
#ifdef Q_OS_MAC
            KeyboardMac::Instance()->sendKeyPress(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_LEFT));
#endif
            return;
        }
        else if (cmd.contains("<right>"))
        {
            readableCommand = "Right";
#ifdef Q_OS_MAC
            KeyboardMac::Instance()->sendKeyPress(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_RIGHT));
#endif
            return;
        }
        else if (cmd.contains("<space>"))
        {
            readableCommand = "Space";
#ifdef Q_OS_MAC
            KeyboardMac::Instance()->sendKeyPress(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_SPACE));
#endif
            return;
        }
        else
        {
            if (Settings::Instance()->serialCommands)
            {
                readableCommand = getCommandValue(cmd);
                Command *command = new Command();
                if (readableCommand.contains("<0>"))
                {
                    command->data->append(char(0));
                }
                else if (readableCommand.contains("<1>"))
                {
                    command->data->append(1);
                }
                else if (readableCommand.contains("<2>"))
                {
                    command->data->append(2);
                }
                else
                {
                    command->data->append(char(0));
                }
                Serial::Instance()->sendCommand(*command);
            }
            return;
        }
    }
    else
    {
        readableCommand  = getCommandValue(cmd);
        if (readableCommand.toLower().contains("execute") && Settings::Instance()->serialCommands)
        {
            Command *command  = new Command();
            if (readableCommand.contains("10"))
            {
                command->data->append(char(0));
            }
            else if (readableCommand.contains("11"))
            {
                command->data->append(1);
            }
            else if (readableCommand.contains("12"))
            {
                command->data->append(2);
            }
            else
            {
                command->data->append(char(0));
            }
            Serial::Instance()->sendCommand(*command);
        }
        else if (readableCommand == " ")
        {
            readableCommand = "Space";
#ifdef Q_OS_MAC
            KeyboardMac::Instance()->sendKeyPress(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_SPACE));
#endif
        }
        else
        {
#ifdef Q_OS_MAC
            KeyboardMac::Instance()->sendEachKey(readableCommand);
#endif
        }
    }
}

void ApiV1::parseMediaCommand(QString cmd)
{
    QString value = getCommandValue(cmd);
    if (value == "play")
    {
        readableCommand = "Play";
        Media::Instance()->playMedia();
    }
    else if (value == "stop")
    {
        readableCommand = "Stop";
        Media::Instance()->stopMedia();
    }
    else if (value == "next")
    {
        readableCommand = "Next";
        Media::Instance()->nextMedia();
    }
    else if (value == "prev")
    {
        readableCommand = "Previous";
        Media::Instance()->previousMedia();
    }
    else if (value == "volup")
    {
        readableCommand = "Volume up";
        Media::Instance()->volumeUp();
    }
    else if (value == "voldown")
    {
        readableCommand = "Volume down";
        Media::Instance()->volumeDown();
    }
    else if (value == "mute")
    {
        readableCommand = "Mute";
        Media::Instance()->volumeMute();
    }
    else if (value == "launch")
    {
        readableCommand = "Launch player";
        Media::Instance()->launchPlayer();
    }
}

void ApiV1::parseScrollCommand(QString cmd)
{
    QString value = getCommandValue(cmd);
    if (value == "zoomin")
    {
        readableCommand = "Zoom in";
        MouseMac::Instance()->zoom(1,1);
    }
    else if (value == "zoomout")
    {
        readableCommand = "Zoom out";
        MouseMac::Instance()->zoom(-1,1);
    }
    else if (value == "back")
    {
        readableCommand = "Back";
#ifdef Q_OS_MAC
        KeyboardMac::Instance()->sendShortcut(KeyboardMac::KEYCODE_BROWSER_BACK);
#endif
    }
    else if (value == "forward")
    {
        readableCommand = "Forward";
#ifdef Q_OS_MAC
        KeyboardMac::Instance()->sendShortcut(KeyboardMac::KEYCODE_BROWSER_FORWARD);
#endif
    }
    else if (value == "pageup")
    {
        readableCommand = "Page up";
#ifdef Q_OS_MAC
        KeyboardMac::Instance()->sendKeyPress(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_PAGE_UP));
#endif
    }
    else if (value == "pagedown")
    {
        readableCommand = "Page down";
#ifdef Q_OS_MAC
        KeyboardMac::Instance()->sendKeyPress(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_PAGE_DOWN));
#endif
    }
    else if (value == "cancel")
    {
        readableCommand = "Cancel";
#ifdef Q_OS_MAC
        KeyboardMac::Instance()->sendShortcut(KeyboardMac::KEYCODE_CANCEL);
#endif
    }
    else if (value == "refresh")
    {
        readableCommand = "Refresh";
#ifdef Q_OS_MAC
        KeyboardMac::Instance()->sendShortcut(KeyboardMac::KEYCODE_REFRESH);
#endif
    }
    else if (value == "fullexit")
    {
        readableCommand = "Exit fullscreen";
#ifdef Q_OS_MAC
        KeyboardMac::Instance()->sendShortcut(KeyboardMac::KEYCODE_FULLSCREEN);
#endif
    }
    else if (value == "fullscreen")
    {
        readableCommand = "Fullscreen";
#ifdef Q_OS_MAC
        KeyboardMac::Instance()->sendShortcut(KeyboardMac::KEYCODE_FULLSCREEN);
#endif
    }

}

void ApiV1::parseShortcutCommand(QString cmd)
{
    QString value = getCommandValue(cmd);
    if (value == "desktop")
    {
        readableCommand = "Show desktop";
#ifdef Q_OS_MAC
        KeyboardMac::Instance()->sendShortcut(KeyboardMac::KEYCODE_SHOW_DESKTOP);
#endif
    }
    else if (value == "close")
    {
        readableCommand = "Close";
#ifdef Q_OS_MAC
        KeyboardMac::Instance()->sendShortcut(KeyboardMac::KEYCODE_CLOSE);
#endif
    }
    else if (value == "copy")
    {
        readableCommand = "Copy";
#ifdef Q_OS_MAC
        KeyboardMac::Instance()->sendShortcut(KeyboardMac::KEYCODE_COPY);
#endif
    }
    else if (value == "paste")
    {
        readableCommand = "Paste";
#ifdef Q_OS_MAC
        KeyboardMac::Instance()->sendShortcut(KeyboardMac::KEYCODE_PASTE);
#endif
    }
    else if (value == "selectall")
    {
        readableCommand = "Select all";
#ifdef Q_OS_MAC
        KeyboardMac::Instance()->sendShortcut(KeyboardMac::KEYCODE_SELECT_ALL);
#endif
    }
    else if (value == "undo")
    {
        readableCommand = "Undo";
#ifdef Q_OS_MAC
        KeyboardMac::Instance()->sendShortcut(KeyboardMac::KEYCODE_UNDO);
#endif
    }
    else if (value == "standby")
    {
        readableCommand = "Standby";
#ifdef Q_OS_MAC
        KeyboardMac::Instance()->standby();
#endif
    }
    else if (value == "shutdown")
    {
        readableCommand = "Shutdown";
#ifdef Q_OS_MAC
        KeyboardMac::Instance()->shutdown();
#endif
    }
}

void ApiV1::parseSlideshowCommand(QString cmd)
{
    QString value = getCommandValue(cmd);
    if (value == "pause")
    {
        readableCommand = "Pause slideshow";
#ifdef Q_OS_MAC
        KeyboardMac::Instance()->sendKeyPress(kVK_ANSI_B);
#endif
    }
    else if (value == "next")
    {
        readableCommand = "Next slide";
#ifdef Q_OS_MAC
        KeyboardMac::Instance()->sendKeyPress(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_RIGHT));
#endif
    }
    else if (value == "prev")
    {
        readableCommand = "Previous slide";
#ifdef Q_OS_MAC
        KeyboardMac::Instance()->sendKeyPress(KeyboardMac::Instance()->keycodeToKey(KeyboardMac::KEYCODE_LEFT));
#endif
    }
}
